### ISSUE TITLE
twistd mail plugin esmtp problem

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4796]
|<img alt="tpratt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| tpratt reported|
|-|-|
|Trac ID|trac#4796|
|Type|defect|
|Created|2011-01-08 02:17:43Z|
|Branch|https://github.com/twisted/twisted/tree/esmtp-hostname-error-4796|

It looks as if f.fArgs is uninitialized if you don't specify a hostname on the command line.

```
    if config['smtp']:
        f = smtpFactory()
        f.context = ctx
        if config['hostname']:
            f.domain = config['hostname']
            f.fArgs = (f.domain,)
        if config['esmtp']:
            f.fArgs = (None, None) + f.fArgs
        s = internet.TCPServer(config['smtp'], f)
        s.setServiceParent(config.service)
```

I guess the solution is an else block that gets the hostname:

```
    if config['smtp']:
        f = smtpFactory()
        f.context = ctx
        if config['hostname']:
            f.domain = config['hostname']
        else:
            f.domain = socket.gethostname()
        f.fArgs = (f.domain,)
        if config['esmtp']:
            f.fArgs = (None, None) + f.fArgs
        s = internet.TCPServer(config['smtp'], f)
        s.setServiceParent(config.service)
```

Attachments:

* [ticket-4796.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9ce/9ce6c42986fb2c6cdc0df7900e39a07952b7cc4b/29086304177f7d4941e9dbea2a074b7b15d54d0c.patch) (1649 bytes) - added by tpratt on 2011-01-13 03:03:56Z - Add the appropriate topfile/ file
* [ticket-4796.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9ce/9ce6c42986fb2c6cdc0df7900e39a07952b7cc4b/f1af660f221e4885de2634dab989cdd2cb9614a2.patch) (1071 bytes) - added by tpratt on 2011-03-08 02:29:35Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4796 4796
type__defect defect
reporter__tpratt tpratt
priority__normal normal
milestone__ 
branch__branches_esmtp_hostname_error_4796 branches/esmtp-hostname-error-4796
branch_author__tpratt__exarkun tpratt, exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1294453063000000 1294453063000000
changetime__1312837115000000 1312837115000000
version__None None
owner__exarkun exarkun
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="tpratt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|tpratt commented|
|-|-|

I should have posted this first:

While trying to test some changes, I ran into this:

```
# twistd mail -E
/usr/lib/python2.6/dist-packages/twisted/mail/smtp.py:10: DeprecationWarning: the MimeWriter module is deprecated; use the email package instead
  import MimeWriter, tempfile, rfc822
Traceback (most recent call last):
  File "/usr/bin/twistd", line 19, in <module>
    run()
  File "/usr/lib/python2.6/dist-packages/twisted/scripts/twistd.py", line 27, in run
    app.run(runApp, ServerOptions)
  File "/usr/lib/python2.6/dist-packages/twisted/application/app.py", line 626, in run
    runApp(config)
  File "/usr/lib/python2.6/dist-packages/twisted/scripts/twistd.py", line 23, in runApp
    _SomeApplicationRunner(config).run()
  File "/usr/lib/python2.6/dist-packages/twisted/application/app.py", line 374, in run
    self.application = self.createOrGetApplication()
  File "/usr/lib/python2.6/dist-packages/twisted/application/app.py", line 434, in createOrGetApplication
    ser = plg.makeService(self.config.subOptions)
  File "/usr/lib/python2.6/dist-packages/twisted/mail/tap.py", line 182, in makeService
    f.fArgs = (None, None) + f.fArgs
AttributeError: ESMTPFactory instance has no attribute 'fArgs'
```

I was running trunk code at the time, but I also tried it on a pristine Ubuntu using Twisted 10.1 from the Ubuntu repo and had the same issue.

**Comment by twisted-trac** [comment]
|<img alt="tpratt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tpratt removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to tpratt|
|-|-|

Seems more or less the right approach.  A couple things:

  1. The default for `domain` is a little complicated.  See how `twisted.mail.smtp.DNSNAME` is initialized.  It might be better to leave the default as `None` and avoid setting the attribute if it is not specified on the command line.
  1. Needs some unit tests.  Doesn't have to be anything too complicated, just exercise the code enough that you see that you make it past the part where it blows up without this fix, and maybe verify that the right type comes back from `makeService`, or that some of its attributes have the right value, stuff like that.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="tpratt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|tpratt removed owner|
|-|-|

Simpler solution as recommended.

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_6"></a>|@moshez set owner to tpratt|
|-|-|

The test docstring is incorrect, it seems?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [32421]) Branching to 'esmtp-hostname-error-4796'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [32422]) Apply tpratt's patch with a more complete unit test added

refs [#4796](https://github.com/twisted/twisted/issues/4796)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun removed owner|
|-|-|

Fixed the docstring and added another assert to the test method, code is in the branch, [build results are here](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/esmtp-hostname-error-4796).

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_11"></a>|@jesstess set owner to @exarkun|
|-|-|

Thanks for reporting and creating a patch for this, tpratt, and thanks for seeing this through, exarkun.

The intention of the patch looks good. However, ```test_esmtpWithoutHostname``` says ```makeService``` raises a ```UsageError```, but ```makeService``` isn't invoked in this test, so it should probably either move out of ```MakeServiceTests``` and into ```OptionsTestCase``` and get an updated docstring, or get tweaked to test this options code path through ```makeService```.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

Argh.  Thanks for the review, sorry about the silly mistake.  I [changeset:32437 moved the test into `OptionsTestCase`], where it should be, since the exception should be raised before `makeService` is ever called.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_13"></a>|@jesstess set owner to @exarkun|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [32438]) Merge esmtp-hostname-error-4796 

Author: tpratt, exarkun
Reviewer: jesstess
Fixes: [#4796](https://github.com/twisted/twisted/issues/4796)

Check for `twistd mail`'s `--hostname` option whenever the `--esmtp` option has
been given and emit a useful error message if the former is not found, as the latter
requires it.
